src/grocy.rs

```diff
@@ -297,3 +322,39 @@ fn prompt_location(grocy_state: &GrocyState, default_location: u32) -> Result<Lo
 
     Ok(location)
 }
+
+fn prompt_barcode_userfields(grocy_api: &GrocyApi) -> Result<HashMap<String, String>> {
+    let userfields = grocy_api.get_barcode_userfields()?;
+    let mut values = HashMap::new();
+
+    for field in userfields {
+        let msg = format!("{} for this barcode:", field.caption);
+        let options = field
+            .config
+            .as_deref()
+            .map(|config| config.lines().map(|l| l.to_owned()).collect());
+
+        let prompt: Option<Box<dyn Prompt>> = match field.r#type.as_str() {
+            "checkbox" => Some(Box::new(Confirm::new(&msg))),
+            "text-single-line" => Some(Box::new(Text::new(&msg))),
+            "date" => Some(Box::new(DateSelect::new(&msg))),
+            "preset-list" if options.is_some() => {
+                Some(Box::new(Select::new(&msg, options.unwrap())))
+            }
+            "preset-checklist" if options.is_some() => {
+                Some(Box::new(MultiSelect::new(&msg, options.unwrap())))
+            }
+            _ => None, // unsupported type
+        };
+
+        if let Some(prompt) = prompt {
+            if field.input_required {
+                values.insert(field.name, prompt.prompt()?);
+            } else if let Some(value) = prompt.prompt_skippable()? {
+                values.insert(field.name, value);
```

Comment:
This doesn't work... 
```
error[E0161]: cannot move a value of type `dyn dynprompt::Prompt`
   --> src/grocy.rs:352:43
    |
352 |                 values.insert(field.name, prompt.prompt()?);
    |                                           ^^^^^^ the size of `dyn dynprompt::Prompt` cannot be statically determined

```

Reply:
https://stackoverflow.com/questions/76035679/cannot-move-a-value-of-dyn-type

Use generics instead.

```diff
diff --git a/src/grocy.rs b/src/grocy.rs
index 9cc8d2b..6d2904f 100644
--- a/src/grocy.rs
+++ b/src/grocy.rs
@@ -4,7 +4,7 @@ use std::{collections::HashMap, fmt::Display};
 use anyhow::Result;
 use chrono::{Duration, NaiveDate};
 use grocy::{
-    structs::{Location, ProductDetails},
+    structs::{Location, ProductDetails, UserField},
     GrocyApi,
 };
 use inquire::{Confirm, CustomType, DateSelect, MultiSelect, Select, Text};
@@ -334,27 +334,31 @@ fn prompt_barcode_userfields(grocy_api: &GrocyApi) -> Result<HashMap<String, Str
             .as_deref()
             .map(|config| config.lines().map(|l| l.to_owned()).collect());
 
-        let prompt: Option<Box<dyn Prompt>> = match field.r#type.as_str() {
-            "checkbox" => Some(Box::new(Confirm::new(&msg))),
-            "text-single-line" => Some(Box::new(Text::new(&msg))),
-            "date" => Some(Box::new(DateSelect::new(&msg))),
+        let value: Option<String> = match field.r#type.as_str() {
+            "checkbox" => prompt_userfield(Confirm::new(&msg), &field)?,
+            "text-single-line" => prompt_userfield(Text::new(&msg), &field)?,
+            "date" => prompt_userfield(DateSelect::new(&msg), &field)?,
             "preset-list" if options.is_some() => {
-                Some(Box::new(Select::new(&msg, options.unwrap())))
+                prompt_userfield(Select::new(&msg, options.unwrap()), &field)?
             }
             "preset-checklist" if options.is_some() => {
-                Some(Box::new(MultiSelect::new(&msg, options.unwrap())))
+                prompt_userfield(MultiSelect::new(&msg, options.unwrap()), &field)?
             }
             _ => None, // unsupported type
         };
 
-        if let Some(prompt) = prompt {
-            if field.input_required {
-                values.insert(field.name, prompt.prompt()?);
-            } else if let Some(value) = prompt.prompt_skippable()? {
-                values.insert(field.name, value);
-            }
+        if let Some(value) = value {
+            values.insert(field.name, value);
         }
     }
 
     Ok(values)
 }
+
+fn prompt_userfield<T: Prompt>(prompt: T, field: &UserField) -> Result<Option<String>> {
+    if field.input_required {
+        Ok(Some(prompt.prompt()?))
+    } else {
+        Ok(prompt.prompt_skippable()?)
+    }
+}
```